binnacle/lib/bloc.dart

```diff
@@ -1,56 +1,69 @@
 
+//TODO type these controllers
   final _headingController = BehaviorSubject<String>();
   final _positionController = BehaviorSubject<PositionModel>();
-  final _windContoller = PublishSubject<WindModel>();
-  final _compassController = PublishSubject();
+  final _availableServices = BehaviorSubject();
+  final _activeServices = BehaviorSubject();
+  final _providerData = BehaviorSubject();
+  
+
+  final _windContoller = BehaviorSubject<WindModel>();
+  final _compassController = BehaviorSubject<CompassModel>();
+  final _listAngleController = BehaviorSubject<ListAngleModel>();
 
-  //? Should this contructor be refactored in to an aync factory?
   Bloc() {
-    // * This line is just a dummy position -- delete it when Position works
-    this._positionController.add(PositionModel(lat: 89.4, lon: 121.12));
+
     this._repository = Repository(_positionController);
-    this._positionController.addStream(this._repository.getPositionStream());
+    this._positionController.addStream(_repository.getPositionStream());
     this._windContoller.addStream(_repository.getWindStream());
     this._compassController.addStream(_repository.getCompassStream());
-    
-
-  }
+    this._availableServices.addStream(_repository.getAvailableServices());
+    this._activeServices.addStream(_repository.getActiveServices());
 
+    this._providerData.addStream(_repository.getProviderData());
+    this._listAngleController.addStream(_repository.getListAngleStream());
 
-  Stream<WindModel> get wind => _windContoller;
-  PublishSubject get compass => _compassController.stream; 
-  BehaviorSubject<PositionModel> get position => _positionController.stream; // .stream? 
-
+  }
 
-    // change data
-    //* These don't actually do anything yet. I'm just leaving them 
-    //* as a reference for whene I need to have functions in BLoC
+  BehaviorSubject get availableServices => _availableServices.stream;
+  BehaviorSubject get activeServices => _activeServices.stream;
+  BehaviorSubject get providerData => _providerData.stream;
+  BehaviorSubject<WindModel> get wind => _windContoller.stream;
+  BehaviorSubject<CompassModel> get compass => _compassController.stream;
+  BehaviorSubject<PositionModel> get position =>_positionController.stream;
+  BehaviorSubject<ListAngleModel> get listAngle => _listAngleController.stream;
+  // change data
+  //* These don't actually do anything yet. I'm just leaving them
+  //* as a reference for whene I need to have functions in BLoC
   Function(String) get changeHeading => _headingController.sink.add;
   Function(PositionModel) get changePosition => _positionController.sink.add;
 
-  changeCompassSource() {
-    this._repository.swapCompassStream();
+  setActiveService(ServiceData serviceData) {
+    this._repository.setActiveService(serviceData);
+  }
+  toggleMode(ProviderData providerData) {
+    _repository.toggleMode(providerData);
   }
 
-  
   void dispose() async {
     await _headingController?.drain();
```

Comment:
This is one for Donovan

Reply:
I don't think there's any harm leaving these as optional. Consider the case where you run the app in the emulator or with a mock service which doesn't emit any events. I don't think it should throw an error since that's expected behavior. Also, dispose() would only be called (in theory, it never actually does) when the bloc needs to be torn down. It would be evident before then that they were null